akka-http-core/src/main/scala/akka/http/scaladsl/settings/ConnectionPoolSettings.scala

```diff
@@ -76,6 +96,34 @@ abstract class ConnectionPoolSettings extends js.ConnectionPoolSettings { self:
 }
 
 object ConnectionPoolSettings extends SettingsCompanion[ConnectionPoolSettings] {
+
   override def apply(config: Config) = ConnectionPoolSettingsImpl(config)
+
+  private[akka] def forDefault(system: ActorSystem): ConnectionPoolSettings = {
```

Comment:
Let's make it public and call it `withOverrides`.

Reply:
I think we had some back-and-forth on this one. The issue with adding a `withOverrides` is it becomes ambiguous what takes precedence when you both update individual settings from code and update settings for a how with `withOverrides`.

In its current state, we resolve this ambiguity by making `forDefault`/`withOverrides` private, so when creating a `ConnectionPoolSettings` from code it never takes into account overrides and always chooses the 'global default' settings overridden with the settings from code.

Reply:
> I think we had some back-and-forth on this one. The issue with adding a `withOverrides` is it becomes ambiguous what takes precedence when you both update individual settings from code and update settings for a how with `withOverrides`.

I suspected that but didn't expand enough of the hidden communication in this PR. What I had not understood so far is the significance of doing the overriding not in our code but using config, so that `ConnectionPoolSettings.withOverrides().withMaxConnections(15)` would not apply at all if there's an override block for the target host in the config (because the settings for that host would be copied from the default config in the file).

That's bad indeed. But I also don't like the current solution which introduces another special case to ignore some parts of the config (the overrides) when the user creates a settings object.

Do we have any alternative ideas how to fix that problem? One idea would be to change all the `ConnectionPoolSettings.withXYZ` methods to also apply the change to the configured overrides but I guess that's the case of the ambiguity where it's unclear whether that manual change should actually override the overrides or just set a new default if no override was specified (which isn't represented in our code).

Reply:
Say we have:
```
ConnectionPoolSettings.withOverrides()
  .withMaxConnections(42)
```
And there is a custom configuration configured for `foo.com`, then we'd expect to see `42` used when these `Settings` are used for `foo.com` as well, right? Doesn't that mean `withMaxConnections` would have to be adapted so the new value is applied not only to the global defaults, but also to all `Settings` in the `hostMap`?

Reply:
(ah didn't see your last message before writing mine, these overlap nicely)

Reply:
I think it's reasonable to assume if you do `.withMaxConnections(42)` from code, this is the value you want to use in all cases when using this `Settings` object, regardless of earlier configuration.

I see 2 ways to achieve that behavior:
* adapting all `withXxx` methods to also apply the new setting to all host-specific overrides
* removing `hostMap` from the `Settings`, and solving this earlier: we could introduce a `ConnectionPoolSettings.forHost(hostname)` which would get the per-host defaults for this hostname (with a fallback to the configured defaults).

The advantage of the former approach is it's fairly straightforward to do, but not very elegant.

The advantage of the latter approach is it seems a bit clearer. The disadvantage is it might be hard to implement in a backwards-compatible way, and it's possible able to use a `Settings` for host `X` when actually connecting to host `Y`.

WDYT?

Reply:
> Doesn't that mean `withMaxConnections` would have to be adapted so the new value is applied not only to the global defaults, but also to all `ConnectionPoolSettings` instances in the `hostMap`?

Then the order of changing default settings and override would matter.

I.e.

```
ConnectionPoolSettings.withOverrides()
  .withMaxConnections(42)
  .changeSettingsForHost("example.com") { settings =>
     settings.withMaxConnections(51)
  }
```

would first set all values for `withMaxConnections` to 42 and then again change it only for "example.com". But if you would swap those around, `withMaxConnections` would override the setting for "example.com".

I guess we could accept that. The good thing would be that we don't need the special loading case.

Reply:
So what was the outcome here? Personally I think exposing the overrides in code adds a significant amount of complexity for very marginal benefit (imo). Not just implementation complexity, since that can easily be hidden from users. But just complexity in what it's supposed to do, how to use it, etc.